pkg/activator/net/revision_backends.go

```diff
@@ -0,0 +1,414 @@
+	logger *zap.SugaredLogger, probeFrequency time.Duration) *RevisionBackendsManager {
+	rbm := &RevisionBackendsManager{
+		revisionLister:   revisionLister,
+		serviceLister:    serviceLister,
+		revisionWatchers: make(map[types.NamespacedName]*revisionWatcherCh),
+		updateCh:         updateCh,
+		transport:        transport,
+		logger:           logger,
+		probeFrequency:   probeFrequency,
+	}
+
+	endpointsInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: reconciler.ChainFilterFuncs(
+			reconciler.LabelExistsFilterFunc(serving.RevisionUID),
+			// We are only interested in the private services, since that is
+			// what is populated by the actual revision backends.
+			reconciler.LabelFilterFunc(networking.ServiceTypeKey, string(networking.ServiceTypePrivate), false),
+		),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    rbm.endpointsUpdated,
+			UpdateFunc: controller.PassNew(rbm.endpointsUpdated),
+			DeleteFunc: rbm.endpointsDeleted,
+		},
+	})
+
+	return rbm
+}
+
+func NewRevisionBackendsManager(updateCh chan<- *RevisionDestsUpdate,
+	transport http.RoundTripper, revisionLister servinglisters.RevisionLister,
+	serviceLister corev1listers.ServiceLister, endpointsInformer corev1informers.EndpointsInformer,
+	logger *zap.SugaredLogger) *RevisionBackendsManager {
+	return NewRevisionBackendsManagerWithProbeFrequency(updateCh, transport, revisionLister, serviceLister,
+		endpointsInformer, logger, probeFrequency)
+}
+
+func (rbm *RevisionBackendsManager) getRevisionProtocol(revID types.NamespacedName) (networking.ProtocolType, error) {
+	revision, err := rbm.revisionLister.Revisions(revID.Namespace).Get(revID.Name)
+	if err != nil {
+		return "", err
+	}
+	return revision.GetProtocol(), nil
+}
+
+func (rbm *RevisionBackendsManager) getOrCreateDestsCh(rev types.NamespacedName) (chan []string, error) {
+	rbm.revisionWatchersMux.Lock()
+	defer rbm.revisionWatchersMux.Unlock()
+
+	rwCh, ok := rbm.revisionWatchers[rev]
+	if !ok {
+		proto, err := rbm.getRevisionProtocol(rev)
+		if err != nil {
+			return nil, err
+		}
+
+		destsCh := make(chan []string)
+		rw := newRevisionWatcher(rev, proto, rbm.updateCh, destsCh, rbm.transport, rbm.serviceLister, rbm.logger)
+		rbm.revisionWatchers[rev] = &revisionWatcherCh{rw, destsCh}
+		go rw.run(rbm.probeFrequency)
+		return destsCh, nil
+	}
+
+	return rwCh.ch, nil
+}
+
+// endpointsUpdated is a handler function to be used by the Endpoints informer.
+// It updates the endpoints in the RevisionBackendsManager if the hosts changed
+func (rbm *RevisionBackendsManager) endpointsUpdated(newObj interface{}) {
+	endpoints := newObj.(*corev1.Endpoints)
+	revID := types.NamespacedName{endpoints.Namespace, endpoints.Labels[serving.RevisionLabelKey]}
+
+	destsCh, err := rbm.getOrCreateDestsCh(revID)
+	if err != nil {
+		rbm.logger.Errorw(fmt.Sprintf("Failed to get revision watcher for revision %q", revID.String()),
```

Comment:
To avoid formatting you can just : `logger.Errorw("Failed...n " + revID.String(), zap...)`.

Reply:
That would remove quoting around the revision name, IMO its nice to have that grouping.